modules/phase_field/src/action/MultiAuxVariablesAction.C

```diff
@@ -7,61 +7,72 @@
 
 MultiAuxVariablesAction::MultiAuxVariablesAction(InputParameters params) :
-    AddAuxVariableAction(params)
+    AddAuxVariableAction(params),
+    _grain_num(getParam<unsigned int>("grain_num")),
+    _var_name_base(getParam<std::vector<std::string> >("variable_base")),
+    _num_var(_var_name_base.size()),
+    _data_type(getParam<MultiMooseEnum>("data_type")),
+    _data_size(_data_type.size())
 {
 }
 
-/*
-The name of the variable is the variable name base followed by two digits.
-The first digit in the name is the dimension it applies to:
-  0 = x, 1 = y, 2 = z (or some other criteria as user needs).
-The second digit in the name is the order parameter it applies to.
-*/
-
 void
 MultiAuxVariablesAction::act()
 {
-  MooseEnum order = getParam<MooseEnum>("order");
-  MooseEnum family = getParam<MooseEnum>("family");
-  unsigned int op_num = getParam<unsigned int>("op_num");
-  std::vector<std::string> var_name_base = getParam<std::vector<std::string> >("var_name_base");
-  std::set<SubdomainID> blocks = getSubdomainIDs();
-  unsigned int dim = getParam<unsigned int>("dim");
+  if (_num_var != _data_size)
+    mooseError("Data type not provided for all the auxvariables in MultiAuxVariablesAction");
 
-  unsigned int size_v = var_name_base.size();
+  // Blocks from the input
+  std::set<SubdomainID> blocks = getSubdomainIDs();
+  // mesh dimension required for gradient components
+  unsigned int dim = _mesh->dimension();
 
   // Loop through the number of order parameters
-  for (unsigned int val = 0; val < size_v; ++val)
-    for (unsigned int op = 0; op < op_num; ++op)
-      for (unsigned int x = 0; x < dim; ++x)
+  for (unsigned int val = 0; val < _num_var; ++val)
+    for (unsigned int gr = 0; gr < _grain_num; ++gr)
+    {
+      /// for exatrcting data from MaterialProperty<std::vector<Real> >
+      if (_data_type[val] == "Real")
       {
-        //Create variable names
-        std::string var_name = var_name_base[val] + Moose::stringify(x) + Moose::stringify(op);
+        //Create variable names with variable name base followed by the order parameter it applies to.
+        std::string var_name = _var_name_base[val] + Moose::stringify(gr);
+
         if (blocks.empty())
           _problem->addAuxVariable(var_name, _fe_type);
         else
           _problem->addAuxVariable(var_name, _fe_type, &blocks);
       }
+      /// for exatrcting data from MaterialProperty<std::vector<RealGradient> >
+      if (_data_type[val] == "RealGradient")
+        for (unsigned int x = 0; x < dim; ++x)
+        {
+          /*
+          / The name of the variable is the variable name base followed by two digits.
+          / The first digit in the name is the order parameter it applies to.
+          / The second digit in the name is the dimension it applies to:
+          / 0 = x, 1 = y, 2 = z (or some other criteria as user needs).
+          */
+          std::string var_name = _var_name_base[val] + Moose::stringify(gr) + Moose::stringify(x);
```

Comment:
I think it would be better to have `_x`, `_y`, and `_z` as suffixes (to improve the experience with visualization software like Paraview)